mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py

```diff
@@ -0,0 +1,163 @@
+            documents_url_reader = SimpleWebPageReader(html_to_text=True).load_data([args['using']['source_url_link']])
+            pred = documents_url_reader 
+
+        else:
+            raise Exception(f"Invalid operation mode. Please use one of {self.supported_reader}.")  
+        
+        self.model_storage.file_set('pred', dill.dumps(pred))
+        self.model_storage.json_set('args', args)
+
+        
+
+    def predict(self, df: Optional[pd.DataFrame] = None, args: Optional[Dict] = None) -> None:
+
+        args = self.model_storage.json_get('args')
+        data_docs = dill.loads(self.model_storage.file_get('pred'))
+        input_keys = list(args.keys())
+
+        input_column = args['using']['input_column']
+
+        if input_column not in df.columns:
+            raise RuntimeError(f'Column "{input_column}" not found in input data')
+
+        
+        query_engine  = self.predict_qa_reader(data_docs)
+                
+            
+        questions = df[input_column]   
+        results = []
+        
+        
+        for question in questions:
+            query_results = query_engine.query(question)
+            results.append(query_results)
+        
+        result_df = pd.DataFrame({'question': questions, 'predictions': results})
+
+        result_df = result_df.rename(columns={'predictions': args['target']})
+
+        return result_df
+
+
+    def _get_llama_index_api_key(self, args, strict=True):
+        """ 
+        API_KEY preference order:
+            1. provided at model creation
+            2. provided at engine creation
+            3. OPENAI_API_KEY env variable
+            4. llama_index.OPENAI_API_KEY setting in config.json
+        """ 
+        # 1
+        if 'OPENAI_API_KEY' in args:
+            return args['OPENAI_API_KEY']
+        # 2
+        connection_args = self.engine_storage.get_connection_args()
+        if 'OPENAI_API_KEY' in connection_args:
+            return connection_args['OPENAI_API_KEY']
+        # 3
+        api_key = os.getenv('OPENAI_API_KEY')
+        if api_key is not None:
+            return api_key
+        # 4
+        config = Config()
+        openai_cfg = config.get('llama_index', {})
+        if 'OPENAI_API_KEY' in openai_cfg:
+            return openai_cfg['OPENAI_API_KEY']
+
+        if strict:
+            raise Exception(f'Missing API key "OPENAI_API_KEY". Either re-create this ML_ENGINE specifying the `OPENAI_API_KEY` parameter,\
+                 or re-create this model and pass the API key with `USING` syntax.')  
+
+    def predict_qa_reader(self,documents):
+        """ 
+        connects with llama_index python client to predict the q&a
+
+        """ 
+        
+        args = self.model_storage.json_get('args')
+
+        os.environ['OPENAI_API_KEY'] = args['using']['openai_api_key']
+        
+        if args['using']['index_class'] == 'GPTVectorStoreIndex':
+            index = GPTVectorStoreIndex.from_documents(documents)
```

Comment:
this is for a future PR - we actually have a `GPTPandasIndex` class that can operate directly over a pandas dataframe (and querying the index would just create a sequence of pandas operations that run in a python REPL) 

could be better than using a vector store (though may need to think about security)